### ISSUE TITLE
Automated tests for quickfort and its dependencies

### ISSUE THREAD

**Issue by myk002** [DFHack/dfhack#1690]
Long term, what I’m envisioning is running quickfort tests as part of the automated regression suite so that the tests prevent changes from being merged if they break any existing functionality.
Short term, I just want an easy way to discover bugs and unhandled edge cases in quickfort and its dependencies, and to make sure *I* don’t break anything as I make changes in quickfort..

So I’m thinking this project can be broken down to two phases:
1. Infrastructure and code for manually-initiated locally-run quickfort regression tests
1. Integrate quickfort tests into DFHack’s automated regression testing

In the first phase, I’d like to:
- Enable fast iteration during dev (must at least run on Linux) by running tests interactively on an already-loaded map
- Collocate test files with code files (same directory, or at least same repo) so tests can be reviewed and merged with the code that they test
- Provide Lua code coverage visualization
- Achieve at least 75% code coverage for quickfort

And in the second phase, I’d like to:
- Gate code submission (must show a pass on GitHub before PR can be merged)
- Run on all platforms

## Questions to answer for phase 1
- What should the interface be for the developer to run the tests? What manual setup should be required?
- Can we easily use the existing test harness in DFHack/dfhack to run the tests? If not, what should we use?
- How do we get coverage reports?
- How do we structure the tests so that the code is clean but local functions can be accessed for testing?
- Is it ok to add (tens of?) thousands of lines of test code to DFHack/scripts? Can/should they be included in the distribution? What about test blueprints?
- How can the work we do in this phase positively impact the rest of DFHack? Can we make writing and running tests easier for contributing developers?

Here are my initial thoughts (as a starting point for discussion):

### Interface

The primary use case in this phase will be me writing and executing tests while actively developing for a PR. Quick edit-save-run cycles. It should take no more than a few seconds to run tests, and it should be really easy to rerun the most recently-run test suite.

The developer would be required to load up a fort within which the tests will run. Test set-up functions can forbid all existing items and call tiletypes and createitem to normalize the environment.

The test command, including filters and test output parameters, would be run from the DFHack# prompt. I suspect it would be best to implement the command within quickfort for now: `quickfort test …` while we figure out the test command feature set. If we want to make this into a framework for all scripts, we can move the logic into a common location later.

### Test harness selection

The test harness in DFHack/dfhack/test/ is not an obvious fit for the ad-hoc interactive test running of phase 1. It has the following shortcomings (as seen from the phase 1 use case):
- Configuration is specified in a file instead of command line arguments
- The test harness starts each test run from the DF title screen, which would seriously slow down test iteration during active development
- Test files for the scripts repo have to be added to a directory in the dfhack repo. This doubles the PR load.
- It doesn’t support running tests on a live fortress map, which will be required for quickfort integration tests

And although we could work to fix these shortcomings (the third item in particular will be a hard blocker for phase 2), it seems like there might be a way to keep DFHack/dfhack/test focused on its current role of running full verification suites, and use a different test harness for the actual quickfort tests that is better suited for interactivity.

I haven’t researched every Lua testing framework out there, but [LuaUnit](https://github.com/bluebird75/luaunit) seems well-regarded, maintained, and acceptably-licensed (BSD). However, it is also not perfect.

Pros:
- rich assertion library
- programmatically configurable filtering
- nice readable test summary output
- flexible format for tests -- put them anywhere and they get autodiscovered

Cons:
- one more dependency
- duplicates some functionality that DFHack already has (test suite management, assertions, pretty printing)
- uses _G to scan for test functions (can work around by accessing lower-level functions directly)
- calls os.exit (can work around by calling lower-level functions or by pre-checking params)

### Coverage reports

We can use [luacov](https://github.com/keplerproject/luacov) (MIT license) with [Cluacov](https://github.com/mpeterv/cluacov) plugins (MIT license) to measure coverage and generate reports. Luacov can just be directly installed to the DFHack library dir. Cluacov would need to be built. Need DFHack options that dev can specify to load luacov when the lua interpreter is created and specify coverage checking filters. Luacov needs to be the first thing loaded, so we can't just add it to the scripts (I think).

### Test structure

Writing the tests inside the code files would keep them close to the logic that they’re testing, but the code files would become impossibly long. In my experience test code tends to outsize real business logic by about 3 to 1. That would make the code difficult to read. How else to access local functions for unit testing, though?

I’m thinking of creating a (global) table in each file named test_hooks with references to testable functions, local and non-local. The tests can then live in a separate file and they can access target functions via the test_hooks table.

We currently ‘require’ all internal/ files in the main quickfort.lua script to refresh their environments if the files have changed, but we'd only 'require' the test files if the 'test' command is being run. This would keep potentially large test files from impacting normal usage of quickfort.

I think we should split unit tests and integration tests into different files. They are likely to be written in very different styles and have different dependencies. For example, unit tests might not even need to have a map loaded. It would be useful to measure coverage of the two types of tests separately as well.

internal/quickfort/foo-unit-tests.lua, foo-integration-tests.lua
internal/quickfort/test-blueprints/*.csv,*.xlsx

### Should the test code be part of the distribution?

Not my decision, but some thoughts:

Pros:
- No packaging changes required to filter them out
- Blueprint authors can use the test blueprints as examples
- Easy for potential new developers to examine and run the tests

Cons:
- Increases size of distribution (unsure by how much. probably a few hundred K)
- Messy?

### Scaling impact

The first benefactor of this work will be quickfort (and dependencies), but we will create infrastructure that can be leveraged by other scripts. Would contributing developers ever write tests? I don't know. Probably not without testing requirements mandated by the DFHack core team. We can make testing possible, though. Generalize the interface and document it. See who uses it.

## Questions to answer for phase 2
- How should we implement fortress mode in the test harness?
- How do we call into the tests written in phase 1?

### How to implement MODE_NAVIGATE_FNS.fortress

More questions than answers here. Do we write the code to gin up a new region for every test, select an embark site, and embark with some fresh dwarves? Would it be acceptable to manually pre-create a save for the current DF version, download it on the build server, and just load that? That would be simpler, less prone to flaky failures, and DF versions change slowly enough that it might be an acceptable maintenance burden. 

What about running the tests locally, where a pre-generated save for the build server is not accessible? Require that a dev has an appropriate fort in the first save slot? We can detect if there is something in region1 and error out if not (or just print a warning and skip running fortress-mode tests). We won’t be saving changes to that fort so this might be ok. Since each test suite is responsible for normalizing the world before running tests, it shouldn’t matter what the state of the savegame is.

Should we reload the save for every test? That might be overkill. I suspect it would be sufficient to just load it once and leave it up to the test functions to get the world into a known good state for the tests. They’ll have to do that anyway to make sure tests still pass when the save is regenerated for the next DF version and is slightly different. We *could* reload the save once per test suite so at least a suite can’t be broken by another suite. Is that feasible? How long does a save take to load on the build server?

I'm looking through a quickfort lens here, though, and I might not be thinking generally enough. What other scripts would benefit from running tests in a live map? What requirements would they have on the state of the map? Will other scripts always be able to programmatically create the environment they’d need to run their tests?

### How do we integrate with the phase 1 tests?

We could write a shim that we place in the DFHack/dfhack/test directory that calls the quickfort tests that we implemented in phase 1. The statistics will be off for number of tests run, number of assertions, etc., but the end pass/fail result should be correct, and we'll still be able to get failure details from the luaunit console output.

We could prototype the integration before implementing fortress mode by only running unit tests that don't need a fort map.

### Phase 1: manually-initiated quickfort tests

Goal: detect issues with manually-initiated test runs

1. Luaunit, luacov, and Cluacov are included in DFHack
1. DFHack can load and configure luacov when starting the Lua interpreter
1. `quickfort test` command implemented with skeleton infrastructure, driven by luaunit
1. First unit tests are written, can be run with quickfort test, and coverage report makes sense
1. First integration tests are written, can be run with quickfort test, and coverage report makes sense
1. All new quickfort code required to have tests

### Phase 2: automated regression tests run on every PR

Goal: prevent code that breaks quickfort from being checked in

1. MODE_NAVIGATE_FNS.fortress is implemented in Dfhack test/main.lua
1. Quickfort tests are run from Dfhack test/main.lua using MODE_NAVIGATE_FNS.fortress

**Comment by lethosor** [comment]
> * Configuration is specified in a file instead of command line arguments

In case you weren't aware: you can pass command-line arguments to https://github.com/DFHack/dfhack/blob/develop/travis/run-tests.py, which get written to the config file. I think I did this because Python's `argparse` offered some features that we don't have on the Lua side, and it avoids the need to translate `run-tests.py` arguments to `test/main.lua` arguments.

`test/main.lua` *can* be run manually, though - and if `MODE_NAVIGATE_FNS.fortress` is implemented, running just tests that require a fortress once a fortress is loaded should be relatively fast. You've raised some good points about implementing this well. I think we could come up with a small save in the 1-3MB range that's usable for at least tests that need a map. Supporting tests that require extra, custom setup (e.g. crafting items) would be harder. (In this particular case, `createitem` could be used by those tests, but other tests may have more complicated requirements.)

The current unit tests are not the easiest to set up and use manually, though - they were really intended for use by CI. I would certainly be willing to improve that. I'm curious if `quickfort test` could be made to run `test/main` under the hood - it already supports running tests from a certain folder and filtering them by file/name.

----

Regarding some other points:

I'm also interested in what it would take to support `luacov`. Our entrypoint for scripts is a bit unconventional (lots of Lua code has already run before the first line of a script is reached) but embedding Lua is common enough that I would expect there is a solution. `cluacov` may be harder to integrate, as our build system doesn't handle Lua extensions written in C yet. It may also be tripped up by the fact that we build Lua as C++, not C (I'm not sure how low-level it gets). Another issue with both of these is that Lua only supports registering one debug hook at a time, and core logic to support `kill-lua` already uses a debug hook. We have control over that logic, though, and could probably get (c)luacov to work with it.

I'm a little reluctant to use `luaunit` and the current framework, but I guess I'm not sure what the advantages of luaunit are. If there are clear advantages, I could be persuaded to switch over to it.

I've thought a bit about ways to reload saves before test runs/suites. One idea I had was a plugin capable of saving/restoring the state of important globals, but there's a decent chance that this could cause issues of its own. I haven't had time to experiment with this yet.

**Comment by myk002** [comment]
funny you should mention arg parsing - I have it on my short list to introduce getopt to the scripts. the quickfort command line could use more flexibility. right now `quickfort list dreamfort -l` is valid but (the more natural) `quickfort list -l dreamfort` is not. using getopt would fix this.

I wasn't aware of run-tests.py as an interface to the test harness. That is useful for phase 2, but it would need some work to support the developer workflow -- quick run-edit-rerun tests in an already-loaded map. For example, I really wouldn't want to relaunch df just to rerun a unit test that I'm trying to get passing.

Running `test/main.lua` directly would be more appropriate for the dev workflow. Doing commandline parsing and config file writing in `quickfort test` (essentially porting `run-tests.py` to lua) and then calling `test/main.lua` could be made to work.

I found this example of luacov being used as a VM in a C++ process: https://github.com/keplerproject/luacov/issues/53. We can probably work around the conflicting debug hooks by having `kill-lua force` restore the previous debug hook after it takes control and injects the error. It's probably ok to require `force` in this case since luacov will only be enabled if a developer specifically requests it.

It looks like we can luaL_dofile (or similar) on luacov.lua as one of the first lines in Lua::Open(), maybe after interrupt_init() so it can steal the debug hook. I'll have to see how our file loading mechanisms play with luacov. We might have to absolute-ize path names if luacov has trouble picking them up.

cluacov is optional -- it's just a performance and accuracy enhancer. It says that it reduces the number of lines erroneously marked as missed, but I'm not sure how much of a problem that will be for us. We'll have to see if it's worth spending time on.

as far as i can see, the advantages of `luaunit` are:
- more comprehensive assert library, including recursive table comparisons with useful "diff" output

ok that was a short list. `luaunit` has a lot of features, but most are not relevant for us. It's BSD licensed; maybe we can just adopt the bits we're missing (with attribution, of course).

Reloading the save may or may not become a problem. How much time would we be willing to dedicate to the tests? If we only reload between test suites, how many times is that? O(100)? A 1-3MB save file maybe takes a second to load on a busy buildmaster? 100 seconds doesn't seem too bad, but we'll have a better idea of where the bottlenecks are once we actually get a good number of tests running : p

**Comment by myk002** [comment]
stream of consciousness for my thoughts on implementation. if anything looks strange here then we probably have different assumptions about something and we should poke the ideas until the assumptions fall out.

- pull in luacov as a dependency, install files in <df>/hack/lua/luacov/
- add dfhack flag for initializing luacov
- add luacov initialization to Lua::Open
- make `kill-lua` play nice with the debug hook (profiler.lua can still bork it, though. document the conflict.)
- implement `quickfort test` (document command in the file, but not in the public docs) as test entrypoint; handles config file creation (both test_config.json and [.luacov](https://github.com/keplerproject/luacov#configuration) and calls test/main.lua
- need test_config.json options for skipping straight to running tests in a loaded fort, logic in `main.lua`
- test dir would be scripts/internal/quickfort (for phase 2 would have to change CI test directory to just scripts instead of scripts/test so it picks up everything)
- need to modify `main.lua` so it just picks up files that start with 'test'
- add test/ directory to .gitignore in scripts (I symlink dfhack/scripts to my scripts repo so I can make live changes in a running game, but that means that "installing" the test dir into df_root/hack/scripts makes it end up as an "untracked file" in my scripts repo. is there a better way to do this?)
- add concept of a test suite; setup and teardown functions (I guess this is another advantage that luaunit has)
- see if we can get cluacov working

plus the actual test writing, access via test_hooks table and such.

**Comment by lethosor** [comment]
Sounds good - some additional thoughts from my end:

* 5 seconds for loading a save is probably a better estimate (builds on Buildmaster are often faster than local builds due to caching and the number of cores that it has, not necessarily single-core performance)
* Definitely agree that an easy way to run specific tests in a loaded game is desirable (that's essentially what I meant by `run-tests.py` being designed just for CI purposes - it's not very good for this use-case currently)
* `kill-lua` would likely make a luacov report useless anyway, so requiring `force` to be used and having it overwrite an existing debug hook is fine with me.

> * I symlink dfhack/scripts to my scripts repo so I can make live changes in a running game, but that means that "installing" the test dir into df_root/hack/scripts makes it end up as an "untracked file" in my scripts repo. is there a better way to do this?

[Script paths](https://github.com/DFHack/dfhack/blob/develop/dfhack-config/script-paths.txt) are what I use. I keep forgetting to document this.

I'll try to make some progress on test/main this weekend.

**Comment by myk002** [comment]
Thanks for the script paths pointer! That does make things much cleaner!

Would it be possible to spin the -r4 release before we start on this? Testing might be quite disruptive (refactors and whatnot), and will slow down feature velocity for a while. Quickfort and buildingplan have come a long way since -r3, and I'd hate for all that goodness to be locked away in devel.

**Comment by lethosor** [comment]
Sure - there are a couple other tasks that I want to get done for r4, but not too many, so hopefully it can be ready in a week or so? (Other than your currently-opened PRs, let me know if there's anything else you have planned for r4 that you want to get in)

**Comment by myk002** [comment]
Cool, thanks! Other than the PRs I have open, there's nothing else I have planned for -r4.

**Comment by myk002** [comment]
I've started poking at phase 1 tasks. I'll start submitting PRs once dfhack-0.47.05-r1 is out.

**Comment by myk002** [comment]
I've run into an interesting problem: how can we inform DFHack that we'd like to start with luacov enabled? My original assumption was that we'd "pass a flag" like others (e.g. `--gdb` to start under a debugger) we pass to dfhack-the-script, but we need to control dfhack-the-library, and dfhack-the-library doesn't currently process flags. dfhack-the-library gets its configuration from dfhack.init and other config files, but those are processed too late in the chain to be useful here.

So, what do?

Here are three options I can think of:

1. pass the flag to DF and pull it out of the OS environment (it looks like we can pass unknown flags to DF and it won't choke)
2. make yet another config file that is read earlier in the chain
3. allow the luacov flag to be set after initialization and reinitialize the lua subsystem dynamically

(1) seems fragile to me, since future versions of DF could break it. also, it would need specialized code for each platform, which I don't fancy testing.
(2) is feasible, but I'd only want to do that if we have a good set of configuration that we'd actually want in another config file. new config files need documentation, examples, integration into external tools, etc. there's more potential to confuse users. it doesn't seem worth it for just a single flag. We could simplify this option, though. handle a `--luacov` flag in dfhack-the-script which creates a sentinel file `.luacov.dfhack.enable` which is detected on startup and the existence of the file *is* the flag. the file would need to be deleted by the script if `--luacov` isn't specified. Possible. Feels messy.
So that leaves (3). All scripts would lose state on reinit, but that might not be such a bad thing, especially if we're trying to get coverage of the initialization code. Need to look into how hard this would be to pull off cleanly. However, I'd like to hear your thoughts before I delve too deeply.

**Comment by lethosor** [comment]
Would environment variables be an option? We already use them for some features.

I don't want to do this in the launcher script because it varies across platforms already, and doesn't even exist on Windows.

**Comment by myk002** [comment]
Ah, yes, environment variables are indeed an option. DFHACK_ENABLE_LUACOV?

**Comment by lethosor** [comment]
Fine by me. FWIW, I think the `load-save` script could likely be used for `MODE_NAVIGATE_FNS.fortress` - we would just need a well-defined save to load. Resetting the map might also be possible between tests, although I'm not sure if we'd run into stability issues there.

**Comment by myk002** [comment]
I'm making progress. So far, I've got:

1. luacov added to the build (depends/luacov, pulled in from the [github repo](https://github.com/keplerproject/luacov) unmodified)
2. coverage monitoring working and controllable with an environment variable (DFHACK_ENABLE_LUACOV)
3. env var documentation (Core.rst)
4. report generation script written and documented (devel/luacov.lua)
5. test/main.lua modified to take configuration as commandline params (the commandline I'll be using most often to test will be `test/main.lua --nocache --tests quickfort`. `--nocache` reruns all tests, even if they have a status set in test_status.json)
6. quickfort tests runnable from test/main.lua (place in a top-level `test/` directory in the scripts repo, which merges with the `test/` directory in the dfhack repo upon installation)

I'm starting off with unit testing internal/quickfort/parse.lua since it doesn't affect any game state. I've already found a bug or two (nothing that should block the 0.47.05-r1 release)

**Comment by lethosor** [comment]
> 6\. place in a top-level `test/` directory in the scripts repo, which merges with the `test/` directory in the dfhack repo upon installation

Out of curiosity, what happens if there's a file in both repos with the same name? I wouldn't really expect it to be an issue, but a test file that is overwritten could potentially be hard to notice, since it would just manifest itself as some tests not running. I had considered having tests for scripts live under `test/scripts`, but I'm not sure how inconvenient that would make running an individual script's tests on-demand.

**Comment by myk002** [comment]
Yeah, I agree that it is unlikely, but filename conflicts are absolutely possible. Mandating that script tests live in test/scripts SGTM. Can we enforce with github actions?

The test name filter matches any part of the path, so the extra path component makes no difference. All quickfort tests would be under test/scripts/quickfort/... so I'd run them with just -tquickfort

**Comment by lethosor** [comment]
> Mandating that script tests live in test/scripts SGTM. Can we enforce with github actions?

Well, I meant that files in `scripts/test/` could be installed to `test/scripts/` instead of the top-level `test/`. I don't think it would require additional work in the scripts repo to do that.

**Comment by myk002** [comment]
What are the implications of having the script tests live at different paths during dev and after install? For example, the way I have it set up now is that my dfhack repo and my scripts repo are sibling directories. The scripts repo is +added to dfhack-config/script-paths and my test_config.json sets the test dir to the test/ path in my scripts repo.

In this setup, the in-dev unit tests are never installed. They are run directly from the scripts repo. I don't think this matters for me - substring matching for the test names works the same either way. Are there any use cases where it *does* matter?

**Comment by lethosor** [comment]
Oh, so you're just running a subset of the tests (specifically, tests in the scripts repo)? My installation comment was referring to how all of the tests are installed in `<DF>/tests` when CMake's BUILD_TESTS option is true, which is how tests are run by CI. Ensuring that no tests overwrite each other as part of this process could be useful.

My understanding is that the `--tests-dir` option to `test/main.lua` only supports a single test directory (unless you've changed that). I'm not quite sure what use cases you're asking about, but where tests live shouldn't affect how they run.

**Comment by myk002** [comment]
I'm thinking mostly about the interactive developer experience. I agree that the CI won't care about the paths.

I'm only running a subset of the tests because the tests I care about (the ones I'm actively writing) are not in the installed DFHack dirs. This is fine, and I think working as intended. I just need to document the info developers will need to replicate what I'm doing now.

I've got the cmake changes queued in my local dev repo. I like the organization of putting scripts/test in test/scripts.

**Comment by lethosor** [comment]
> While the tests are in development, they are run as test/testname. once they are installed, they are run as test/scripts/testname.

Ah, that makes sense to me now. Like you said, I wouldn't be too concerned about it. The existing pattern-matching arguments typically wouldn't care about an additional prefix anyway.

I'll try to get r1 up tomorrow, but feel free to put up PRs whenever you like (it's effectively frozen at this point).

**Comment by myk002** [comment]
I created a tiny region for use in fortress-mode tests. I uploaded the save here: https://drive.google.com/file/d/1XvYngl-DFONiZ9SD9OC4B2Ooecu8rPFz/view?usp=sharing

it's ~5MB uncompressed. would that be appropriate to add to the docker image?

**Comment by myk002** [comment]
@BenLubar via discord: I feel like it might be better to put that into dfhack's repository under a folder named testdata or something like that and then have the docker image reproduce the conditions you'd get running the tests straight from the repository without the docker image

@lethosor do you have an opinion on this?

**Comment by lethosor** [comment]
I think I responded on Discord as well, but given that it's larger than most of our binary dependencies that are already hosted [outside of the DFHack repo](https://github.com/dfhack/dfhack-bin), I'd prefer to keep it somewhere else as well. If we ever need to make a new version (whether that's due to DF breaking save compatibility or something else), it would add ~5MB to the repo size every time it's modified, which can add up.

A couple options could include dfhack-bin, DFFD, or files.dfhack.org. Maybe we could put it in multiple places for redundancy.

**Comment by myk002** [comment]
I'm good with most of those options, thoughI think it should go somewhere that requires review by the core team before it can be modified, definitely not dffd.
